### DISCUSSION TITLE
Feature Request: Ability to remove messages from chat history #85

### DISCUSSION THREAD

**Discussion by qpd-v** [cline/cline#888] [Feature Requests]
This will help save on token costs if we can somehow delete messages we don't need from the chats. Especially images!

**Comment by NightTrek** [comment]
Seems like the primary issue with this is prompt caching. Prompt caching gets broken if you delete messages middle out

**Comment by rhokstar** [comment]
Perplexity has this feature. When a user's message/query is removed, the LLM's response is also removed.

Related (also Perplexity feature): If a user's message/query is modified, the LLM's response is regenerated.

**Comment by jpsala** [comment]
This feature is also implemented in cursor and TypingMind, I think it's instrumental, only one unwanted message can add a lot of information to the chat

**Comment by gaieges** [comment]
This would be especially useful, in cases where something like a `rg` tool call results in a TON of content that doesn't need to make it to the LLM
